### DISCUSSION TITLE
Modeller slows down and hangs after some time

### DISCUSSION THREAD

**Discussion by rsoika** [eclipse-glsp/glsp#1089] [General] [answered]
Hi,
I observe a strange behaviour when working on large models with our [Open-BPMN Modeller](https://github.com/imixs/open-bpmn). Such models can have more than 150 elements and after some time (15 minutes) the modeller slows down and hangs in VS-Code. 
Wen I save the model, close it, and open it once again, everything is fine again. I don't think it is related to VS-Code but I observe this as I mostly model in VS-Code.

I have a suspicion that it may be the internal version management and that too many versions of the rather large GLSP model will eventually take up too much memory. Can this be the reason for such a behaviour?

**Comment by planger** [comment]
Hi,

hm, this is a very interesting observation. I haven't seen it myself yet, but I can't rule out that there might be some memory leak or inefficiency in the VS Code integration somewhere, e.g. in the VSCode custom document integration. Can you reproduce this also with the minimal example?

Thank you!

**Comment by rsoika** [comment]
Hi  @planger, 
I did some more tests and it seems to be the VS-Code integration only. I can't reproduce the same error if I run my java server and the Theia Client. In this setup also large models are not consuming more memory. So it seems not to be a GLSP Core issue.

But in VS-Code the problem can be reproduced with Open-BPMN VS-Code Extension in minutes. 
Working in a large model, the editor window consumes more and more memory as more you are working in the model. It is sufficient to just move a single element in a large model around on the plane. In this way you can increase the memory usage from 500MB to 2GB easily in 5 minutes. At the moment when you close just the Modeler-Editor Window in VS-Code the memory is free again. 

Do you have a VSIX file of the Minimal Example somewhere I can use for testing?

**Reply by planger** [reply]
Thank you for investigating this further! I think we should create a Github issue for this topic. Can you please open an issue and report your findings there?

Tobias recently added yarn scripts to create a VSIX file for the workflow diagram VS Code extension:
https://github.com/eclipse-glsp/glsp-vscode-integration/pull/50
With that it should be rather easy to create a minimal VSIX.

If there are any new findings let's collect them in the Github issue to have all the information at one place.
Thanks again for raising and investigating this issue!

**Comment by rsoika** [comment]
I tried out the latest version of the workflow example vscode-extension. I was NOT able to reproduce the problem with the VSIX extension file.

One difference seems to be that the workflow example uses a Node Server, and Open-BPMN is using a Java Server component.
Also my extension is based on `"@eclipse-glsp/vscode-integration": "1.1.0-RC10"` so it seems that I am up to date.

But my problem when migrating to 1.1.0-RC10 was that I did not found an example of how to integrate a Java Server and there was a lot of changes in the api. So at the end I did a lot of guessing how to implement the extension.ts :-/

Maybe you can have just a short look at my code [open-bpmn-extension.ts](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/open-bpmn.vscode/extension/src/open-bpmn-extension.ts) if there is something you are not happy with. 

Or maybe you can provide me an example of how the extension.ts should look like if I want to integrate a Java server instead of nodejs server?

**Reply by rsoika** [reply]
I did some more investigation and tried to sort out things. 

1. Running Open-BPMN within the Theia Platform does not show any issue. No memory leak.
2. But running Open-BPMN in VS-Code shows the memory leak.
3. This happens even when I start the java server part separately and just start the VS-Code Extension without the server part.  

My assumption was that I did something evil in my Server Code on Server side (java) or client side (nodejs). But I think I can sort this out now. I can open a complex model in Theia and VS-Code Extension running the same java server. Only the VS-Code Extension window shows the memory leak. 

And the interesting thing is, that it seems also not to be some of the client <-> sever communication. The memory leak occurs even when I simply grab an element and move it around in the diagram plane for a long time. In this scenario - I think - we have no model updates or modify actions.

![Peek 2023-08-30 23-10](https://github.com/eclipse-glsp/glsp/assets/1339124/6835ef5e-c93c-4b89-8b3b-4d9c238c9764)

**Reply by tortmayr** [reply]
>And the interesting thing is, that it seems also not to be some of the client <-> sever communication. The memory leak occurs even when I simply grab an element and move it around in the diagram plane for a long time. In this scenario - I think - we have no model updates or modify actions.

Yes you are right, there should be no client-server communication in this case. I'm wondering whether the memory leak occurs in the webview or the extension code itself. Could you please use the `Developer tools` to further investigate the memory consumption?
e.g: 
![memory_demo](https://github.com/eclipse-glsp/glsp/assets/2311075/9b9cdd20-2669-46ce-8365-6f2fbdd8e55e)

**Reply by rsoika** [reply]
Thanks for the hint with the developer tools. It doesn't seem to be the extension nor the webview. Maybe this is possibly a general VSCode problem under Linux Debian ....??

![Peek 2023-08-31 00-17](https://github.com/eclipse-glsp/glsp/assets/1339124/4319e522-4c21-4174-a806-252955d05472)

**Reply by rsoika** [reply]
Hi @tortmayr , now I did a test on Windows and it shows that also in Windows VS-Code shows the same high memory consumption. So it is not related to the OS. 
As I told you, I also tested the VSIX extension of the workflow Example and I was not able to reproduce the problem. 
Also Open-BPMN in Theia shows not high memory consumption. So it clearly indicates to me that it is the VS-Code Extension causing the problem.
The Workflow Example Extension is using a nodejs server instead of a java .jar file. Can this be an issue? Do you have a Workflow Example VSIX extension using the java server? Then I could do tests with this extension too. 

But my main concern is that I did something wrong in my implementation. I do not have a blueprint for the VS-Code extension using a java .jar file, so I wonder if you are happy with my [open-bpmn-extension.ts](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/open-bpmn.vscode/extension/src/open-bpmn-extension.ts) and my [open-bpmn-editor-provider.ts](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/open-bpmn.vscode/extension/src/open-bpmn-editor-provider.ts). 

I think at the moment we can rule out that problem caused my the webview - right?

**Reply by rsoika** [reply]
I did not find a way to test the vscode-integration workflow example with running the workflow server as an external java process. Running the workflow server external and launching 'Workflow GLSP Example Extension (External GLSP Server)' results in:

```
[JsonrpcGLSPClient] Connection to server is erroring. Shutting down server. (1) [Error: connect ECONNREFUSED 127.0.0.1:5007
	at TCPConnectWrap.afterConnect [as oncomplete] (node:ne…]
extensionHostProcess.js:108
[JsonrpcGLSPClient] Connection to server is erroring. Shutting down server. (1) [Error: connect ECONNREFUSED 127.0.0.1:5007
	at TCPConnectWrap.afterConnect [as oncomplete] (node:ne…]
```

So I can't test the behaviour in such a setup. I think the `workflow-extension.ts` is not prepared for such a setup?

**Reply by tortmayr** [reply]
Running the workflow Java Server externally and using the ` 'Workflow GLSP Example Extension (External GLSP Server)'` config works fine for me in both Windows & Ubuntu. From my testings I did not encounter any memory leak issues with the Workflow Example even for really large models.

What node version are you using? It seems like there is a recurring connection issue for the  VV scode integration when using Node18 or above (https://github.com/eclipse-glsp/glsp/issues/1100)

>hanks for the hint with the developer tools. It doesn't seem to be the extension nor the webview. Maybe this is possibly a general VSCode problem under Linux Debian ....??

I'm pretty that that should not be the case. We are using Ubuntu on our developer machines which is also Debian-based and have not encountered this issue up until now


So let's quickly sum up what we know until now:
- It does not seem to be a general problem, or at least is not easy reproducible with the workflow example
- We can probably rule out the server as a cultprit because the issue also occurs when only doing client-side actions like dragging an element without a release
- The memory consumption of the webview seems to be normal
- So basically we have ruled out webview and server, so in theory the memory leak should occur in the extension process itself...

**Reply by rsoika** [reply]
Thanks for your reply

```
$ node -v
v16.11.0
```
And the `package.json` file hast the following entry

```
  "engines": {
    "node": ">=16.11.0",
    "yarn": ">=1.7.0 <2.x.x"
  },  
```

And yes your summary is correct. I invested hours in additional testing. Removing nearly everything of my Core Code from Open-BPMN does not solve the problem. It must be the vs-code-integration part.

Can you please show me a template how your write this code with RC10 when using a Java server. Maybe my implementation is wrong...?

**Reply by rsoika** [reply]
One more note: The problem also occurs on blank Windows so we should sort out also OS issues. It must be something with my build. But I also tried now all of my older VSIX versions from my git repo. And all of them show the same behaviour :-(

It also does not matter how big the model is. To see the problem you can simply install my latest version of [Open-BPMN from the market place](https://marketplace.visualstudio.com/items?itemName=open-bpmn.open-bpmn-vscode-extension) and load this simple [demo.bpmn](https://github.com/imixs/open-bpmn/blob/master/open-bpmn.glsp-client/workspace/demo.bpmn).

Now when you grab an element and move it around you can see the memory consumption...

**Reply by tortmayr** [reply]
The problem seems to be related to your glsp-client code somehow. I used to VS Code process explorer for further investigations:
![Screenshot from 2023-09-01 10-01-51](https://github.com/eclipse-glsp/glsp/assets/2311075/2f3803f1-d52f-4e16-acf2-992657ca6f21)

So the memory hog is the webview window.

**Reply by tortmayr** [reply]
Sidenote:
You don't even have do drag an element. Just panning the canvas or zooming is enough to fill up the memory

**Reply by rsoika** [reply]
hm.... I still wonder if my [extension implementation](https://github.com/imixs/open-bpmn-vscode-integration/blob/main/open-bpmn.vscode/extension/src/open-bpmn-extension.ts) is correct... I am using a `GlspSocketServerLauncher` and did some guessing in implementation as I did not found a example matching 1.1.0-RC10. Can you please show me your implementation using a Java Server instead of a nodejs server. just to rule out this part of my code...

**Reply by tortmayr** [reply]
We already can rule out that part of your code. If some would be wrong with your extension implementation than the memory leak would occur in the `extensionHost` process and not the webview window. I`m pretty certain that we can rule out that part of the code.
Also my implementation more or less looks exactly the same as yours.

**Reply by rsoika** [reply]
Ok, interesting. So you think, and this can of course be the case, that my glsp client implementation is the problem? Although I can't reproduce the error in Theia?

**Reply by tortmayr** [reply]
It looks like it, yes. Also the fact that it can be reproduced by pure client-side actions like zooming & panning points in that the direction...

**Comment by rsoika** [comment]
Hi @tortmayr , you where (of course) right. The problem was caused by the diagram itself on the client side. It was observable with Theia in Chromium too but not in Firefox Browser.

The reason for the memory leak is this component which I used to display and edit multi-line text:

```javascript
export class MultiLineTextNode extends ForeignObjectElement implements SArgumentable, EditableLabel {
  readonly isMultiLine = true;
  readonly args: Args;
  text = '';

  override set bounds(bounds: Bounds) {
    /* ignore set bounds, always use the parent's bounds */
  }

  override get bounds(): Bounds {
    if (isBoundsAware(this.parent)) {
      return {
        x: this.position.x,
        y: this.position.y,
        width: this.parent.bounds.width,
        height: this.parent.bounds.height
      };
    }
    return Bounds.EMPTY;
  }

  // @ts-expect-error Arguments are set in the element
  override get code(): string {
    if (this.text === '') {
      const textArg = this.args['text'];
      if (typeof textArg === 'string') {
        this.text = textArg;
      }
    }
    return `<pre>${this.text}</pre>`;
  }

  override namespace = 'http://www.w3.org/1999/xhtml';

  get editControlDimension(): Dimension {
    return {
      width: this.bounds.width - 4,
      height: this.bounds.height - 4
    };
  }
}
```

![Peek 2023-09-02 10-29](https://github.com/eclipse-glsp/glsp/assets/1339124/923d97d3-0bb8-48c3-8a81-7052dc2dbc5e)

The code is the origin example from the documentation [here](https://eclipse.dev/glsp/documentation/rendering/#default-views)

Only removing this part solves the problem.

**Comment by rsoika** [comment, marked as answer]
Finally I solved the issue by removing my `ForeignObjectElement` completely from the diagram. This solved at the end two problems:

- The memory leak in VS-Code and Chrome Browser
- The problem that my labels where not printed in the SVG-Export

I also implemented a new Multi-Line-Label-View (very cool) which I will post in a separate thread soon.

**Reply by planger** [reply]
I'm looking forward to seeing your multi-line label view implementation!
Yes, everything that is a HTML-based foreign object will not work in standard SVG viewers as they are not able to render HTML. Just if you open the SVG in a browser those HTML-based foreign objects will show up.
Back then, when we added HTML-based foreign object support, it seemed the easiest way to support flow text in diagrams, instead of having to implement a flow text algorithm based on SVG on our own. If you have a better solution, we'd be happy to use that by default in GLSP instead!
